Object pools and safer/better ID maps

klaussfreire opened the pull request:
This PR implements a bunch of machinery needed to incrementally construct large fully-connected object graphs.

The problem with fully connected object graphs is that they can't be constructed with the current API without holding them fully in memory. If they're big enough, that can be prohibitive.

This PR allows one to construct proxies into a dynamically growing staging area, replacing in-memory data structures with memory-mapped proxies. This releases some memory pressure by allowing the proxies to take the place of memory-intensive objects within the object graph. Finally, the contents of the dynamic buffers can be dumped together with any remaining objects by proper manipulation of id maps.

To make all this truly work, a lot of changes to the id mapping mechanism had to be introduced. An id map class that automatically discards transient objects was introduced. This is used as the default id map when constructing shared buffers now, since it's far safer than the old default. It's less efficient at object deduplication, however. To gain the old efficiency a set of "stable" objects (ones that we know are going to have a unique id for the duration of the packing procedure) can be specified when constructing the id map.

Furthermore, the id map structure was turned a bit more into a white box by exposing a pair of functions to construct ids: `shared_id` and `wrapped_id`. They can be used to pre-populate id maps with manually packed objects.

Additionally, the derivation of identity keys was changed to accomodate proxies in lieu of actual classes. Now proxy classes use their buffer-offset as identity key, whereas commonly used constants use their value, allowing automatic deduplication of commonly occurring primitive values across the board.

Some bugs in the existing id mapping code were fixed as well as fixes to some collections, like the ObjectIdMapper, that arose during testing.

klaussfreire committed:
Add object pools

An interface that allows building objects in writable buffers
incrementally. Currently quite limited, but limitations can
slowly be lifted as we make the machinery more sophisticated

klaussfreire committed:
Add multi-section test

klaussfreire committed:
Return both proxy and position

klaussfreire committed:
Make returned pos globally unique

klaussfreire committed:
Add random access methods

klaussfreire committed:
Strong id map and pack-repack support

klaussfreire committed:
Fix mapped_dict

klaussfreire committed:
Fix pool idmap handling and proxied_dict's key mapping stability

klaussfreire committed:
Limit idmap growth in object pools

klaussfreire committed:
Turn more classes weakreferenceable

klaussfreire committed:
Better idmaps for repacking proxies

klaussfreire committed:
Add a GenericFileMapper

As a helper utility to map objects rather than contiainers

klaussfreire committed:
Changelog entry

klaussfreire committed:
More idmap improvements

Add support for preloading stable items into StrongIdMap

Add support for equality keys, especially for short strings
and singleton objects

klaussfreire committed:
Limit error strings

klaussfreire committed:
A host of fixes

- Fix is_equality_key on wrapped keys
- Properly pass the idmap and implicit_offs to ObjectIdMapper
  during proxied_dict packing
- Properly store relative offsets on proxied_dict items,
  including header and index values
- Store signed offsets in proxied_dict key pointers. The idmap
  might point to places before the ObjectIdMapper, and in order
  to properly deduplicate those references signed pointers
  are necessary. Do manual casting from numpy's unsigned index
  into a signed pointer to avoid complicating the index structure
- Properly resolve generic proxy typecodes when repacking
  generic object proxies into a generic object attribute, by
  looking up any registered base classes
- Don't slice the buffer in proxied_dicts, this breaks back
  references in the key map. Store a reference to the full buffer
  and an offset, adapt code to match
- Test back references

klaussfreire committed:
Some improvements and fixes:

- Allow marking a stable set of object ids that need not be
  reference-tracked (ie: if you already keep strong references
  on a side structure)
- Make stable_hash and proxied_dict support None as hashable key

klaussfreire committed:
Fix ObjectIdMapper.get, add tests

klaussfreire committed:
Disable tests that are too expensive

klaussfreire committed:
Test nonstring keys

klaussfreire committed:
Expose buffer and offset in generic proxies

klaussfreire committed:
Fix idmap handling of wrapped object attributes

Expose the issue in tests

llogiudice commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +112,256 @@ def _likerobuffer(buf):
+    elif is_equality_key(obj):
+        # Try an equality key, use the object itself as key if hashable
+        try:
+            hash(obj)
```
Shouldn't this take into account the fact that python's hash isn't stable ?

klaussfreire replied:
I don't use the value of the hash, I just use that to test whether the object is hashable

llogiudice commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +112,256 @@ def _likerobuffer(buf):
+@cython.inline
+@cython.returns(cython.bint)
+def is_wrapped_key(obj):
+    # singletons and small strings
```
This doesn't seem to handle small strings (Although it should, I think).

klaussfreire replied:
I think that's a copypasto

llogiudice commented on sharedbuffers/pool.py:
Is this buffer going to be shared among processes ? If so, you need to call ```fsync``` periodically to make sure memory state is correctly synchronized.

klaussfreire replied:
Memory is always synchronized. `msync` (not fsync) would be needed to make sure the changes hit the disk. We don't currently require this.

llogiudice commented on sharedbuffers/pool.py:
```diff
@@ -0,0 +1,175 @@
+
+# Default section size is set to 128MB which is a size at which most
+# malloc implementations turn to mmap
+DEFAULT_SECTION_SIZE = 128<<20
```
According to __man mallopt__, this threshold is at 128**KB**. On the other hand, it appears such threshold is dynamic, although it can be controlled with ```mallopt```.

klaussfreire replied:
It's dynamic, and the top limit for how big that dynamic threshold can get is 64MB.

klaussfreire replied:
By default, of course

klaussfreire committed:
Fix inaccurate comments

klaussfreire committed:
Fix hash of big/signed ints

klaussfreire committed:
Revert NamedTemporaryFile in pools

llogiudice commented on sharedbuffers/pool.py:
Wouldn't it be possible to automatically readjust the section size upon failure ? Or does the caller expect the section size not to change ?

klaussfreire replied:
Are you referring to an exponential growth strategy or something like that? It's already quite big at 128MB, I don't see much benefit from that.

I've thought about enlarging the section by truncating it to a bigger size, but the problem with that is that it would invalidate existing buffers.

llogiudice replied:
Well, it doesn't necessarily have to be exponential, I think something like increasing by 50% on each retry should suffice.

The reason I asked about it is that sharedbuffers can already handle huge data structures gracefully by proxies, and this kind of error doesn't seem like a fatal exception, since you are in control of the temporary file and its size. I guess it's a secondary concern, though.

klaussfreire replied:
Yes, but the caller is in control of both the segment size and the preloaded set. If the caller expects to preload something on all sections that fills up the section size it requested, it's already in deep trouble. Remember this goes on every section that is initialized, so if you're preloading that much data, you need to account for that when setting the section size, or you'll have tons of copies of the same huge data set.

So... I guess it's best if the caller knows about this, rather than "handling" it silently. This clearly is something the caller would not expect, or a bigger section size would have been requested.

klaussfreire replied:
For context, this preload set is a set of objects you preload on all sections and their id maps, to be used as reference by the objects you're planning to add to the section. That's why it's unreasonable to expect the preload set to fill the section, it wouldn't leave any room for useful objects (those you add later with .pack) before another section with another copy of the preload set has to be allocated.

klaussfreire committed:
Make the generic file mapper also return the mmap object

Without it, msync and explicit closing are impossible

klaussfreire committed:
Narrow pointers and some fixes in proxied_tuple

- Fixed proxied_tuple index pointer arithmetic
- Introduce a typecode T with narrow pointers

klaussfreire committed:
Remove redundant import, add changelog entry for narrow pointers

klaussfreire commented:
Ping

klaussfreire committed:
Accelerated widmap calls

klaussfreire committed:
Support add pre-packed buffers to object pools

Allows for parallelized packing through multiprocessing under
certain conditions

llogiudice commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +111,257 @@ def _likerobuffer(buf):
+            self.pop(key, None)
+
+    def __len__(self):
+        return len(self.idmap) + len(self.preloaded)
```
is this actually accurate ? based on the iter protocol below, it appears that the actual length would be the length of the intersection between the idmap and the preloaded map. Then again, it's possible that we don't really care about this length being 100% accurate.

klaussfreire replied:
Yeah, it's not accurate, but it's not worth computing the intersection to get an accurate length either, IMO. Depends on what we use the len for, currently, approximate is enough (it's for logging and stuff like that)

klaussfreire committed:
Better allocation strategy for add

klaussfreire committed:
Merge branch 'master' into object_pool

klaussfreire committed:
Fix condition indentation

klaussfreire committed:
Fix typo

klaussfreire committed:
Fix shared_id for proxied_list slices

klaussfreire committed:
Allow packing base types in object pools

klaussfreire committed:
Add base type packing test

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +111,263 @@ def _likerobuffer(buf):
+    pass
+
+@cython.cclass
+class StrongIdMap(object):
```
Can't we move this to another file instead of adding even more classes to mapped_struct.py?

klaussfreire replied:
This is not the PR to do such refactoring. Also, not until we bump the requirements to a modern Cython version so we can do proper refactoring without a performance hit.

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +111,263 @@ def _likerobuffer(buf):
+        for key in self.preloaded:
+            yield key
+        for key in self.idmap:
+            if key in self and key not in self.preloaded:
```
Why checking if `key in self` and not just checking `key in self.objmap`. Becuase the contains will check on on the idmap and we already know that is true.

klaussfreire replied:
It's easier to maintain. Like this, this code is correct no matter what happens in `__contains__`

klaussfreire replied:
Also, the loop isn't atomic, so items can disappear mid-iteration.

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +111,263 @@ def _likerobuffer(buf):
+                return True
+
+        try:
+            self.objmap[key] = obj
```
I might be missing something if this object is added to the objmap, then the `get` is going to raise KeyError because it wasn't added to the `idmap`?

klaussfreire replied:
The objmap just links the lifetimes of the idmap entry to that of the object. It's a way of detecting stale references. `link` doesn't add items to the idmap, it only links entries to objects, if the entry is there.

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +111,263 @@ def _likerobuffer(buf):
+    if isinstance(obj, tuple):
+        return len(obj) == 2 and obj[0] is WRAPPED
+    elif isinstance(obj, (int, long)):
+        return (obj >> 64) == 0xF
```
Why doing this transformations for int/long? Just to check on `get_wrapped_key`?

klaussfreire replied:
What transformation?

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -315,8 +609,14 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
+            elif dcode == 'T':
+                dtype = 'i'
+            else:
+                raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
```
You are already checking that dcode is `t` or `T`. Why adding this extra exception?

klaussfreire replied:
Just in case. Code changes.

tzulberti-jampp commented on sharedbuffers/pool.py:
```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
```
Aren't you going to cythinize this file?

klaussfreire replied:
I don't see the need for it ATM

gefarion commented on sharedbuffers/mapped_struct.py:
```diff
@@ -106,8 +111,263 @@ def _likerobuffer(buf):
+        rv = (id(lobj.buf) << 64) | lobj.offs
+        if lobj.elem_step != 0:
+            # Add slice arguments to the shared_id
+            rv |= long(lobj.elem_step) << (68 + 64)
```
Why 68 bits? It is not 64 + 64?

klaussfreire replied:
64 + 4 bits for flags

klaussfreire commented:
Only @tzulberti-jampp remaining...

tzulberti-jampp commented on sharedbuffers/pool.py:
I wasn't able to find an example of this function call, so I am not sure about this, but should be that pos == self.write_pos? Why checking the new_end?

klaussfreire replied:
I meant to use it, but I didn't. The intention was to be able to "pack_into" the mmap directly, allocating a generous buffer, and then resize that buffer. But as you say, the old_end should be the one being compared, and I don't have it in this context. Since we didn't use it in the end, because packing directly to the mmap resulted impossible due to mmap module limitations, I'm going to scrap the whole method and call it a day.

tzulberti-jampp commented on sharedbuffers/pool.py:
Is this variable being used?

klaussfreire replied:
Nope

klaussfreire committed:
Remove unused stuff
